mfs/dir.go

```diff
@@ -168,15 +170,51 @@ func (d *Directory) childUnsync(name string) (FSNode, error) {
 	return nil, os.ErrNotExist
 }
 
-func (d *Directory) List() []string {
+type NodeListing struct {
+	Name string
+	Type int
+	Size int64
+	Hash string
+}
+
+func (d *Directory) List() ([]NodeListing, error) {
 	d.lock.Lock()
 	defer d.lock.Unlock()
 
-	var out []string
-	for _, lnk := range d.node.Links {
-		out = append(out, lnk.Name)
+	var out []NodeListing
+	for _, l := range d.node.Links {
+		child := NodeListing{}
+		child.Name = l.Name
+
+		c, err := d.childUnsync(l.Name)
+		if err != nil {
+			return nil, err
+		}
+
+		child.Type = int(c.Type())
+		if c, ok := c.(*File); ok {
+			size, err := c.Size()
+			if err != nil {
+				return nil, err
+			}
+			child.Size = size
+		}
+		nd, err := c.GetNode()
+		if err != nil {
+			return nil, err
+		}
+
+		k, err := nd.Key()
+		if err != nil {
+			return nil, err
+		}
+
+		child.Hash = k.B58String()
+
+		out = append(out, child)
```

Comment:
maybe once IPLD lands we can make unixfs dirs store the sizes too to avoid having to read the entire next layer in order to `ls`....